Custom ClusterRole for RootSync reconcilers

tomasaschan opened the pull request:
Closes #935.

This introduces a new override for `RootSync`s, which allows customization of which `ClusterRole` to bind the reconciler's service account to. The default is `cluster-admin`, which is the same as what was done before this change. For more details, see the included design document.

tomasaschan commented:
Started working on an implementation, but it turns out it was slightly less straightforward than I'd hoped 😄 The main source of complexity is that Config Sync currently creates a single CRB and adds all root reconciler KSAs as subjects.  We must split this into a CRB per RootSync, because

* they might not target the same role, so we cannot have just one CRB for all of them
* when deleting a RootSync we must be able to clean them up, and at that point [we only know the name of the `RootSync`, not the name of the role it used](https://github.com/GoogleContainerTools/kpt-config-sync/blob/main/pkg/reconcilermanager/controllers/garbage_collector.go#L235), so we cannot have one CRB per referenced role.

In practice, we can probably change [`RootSyncPermissionsName`](https://github.com/GoogleContainerTools/kpt-config-sync/blob/main/pkg/reconcilermanager/controllers/build_names.go#L31) to take the name of the `RootSync` as a parameter, fix all compiler errors, and have a working implementation (with different behavior than before wrt the number of bindings created). We'll then also have to change a few tests that assume that permissions will be managed by adding and removing subjects on a single CRB.

tomasaschan commented:
`gpg` timed out when it tried to sign those last two commits; not sure why. I hope GPG-signed commits isn't a requirement for contributing...

tomasaschan commented on pkg/reconcilermanager/controllers/rootsync_controller.go:
```diff
@@ -544,8 +556,8 @@ func (r *RootSyncReconciler) mapObjectToRootSync(obj client.Object) []reconcile.
 	// because all the generated resources have the prefix.
-	roleBindingName := RootSyncPermissionsName()
-	if !strings.HasPrefix(objRef.Name, core.RootReconcilerPrefix) && objRef.Name != roleBindingName {
+	if !strings.HasPrefix(objRef.Name, core.RootReconcilerPrefix) &&
+		!strings.HasPrefix(objRef.Name, fmt.Sprintf("%s:%s", configsync.GroupName, core.RootReconcilerPrefix)) {
```
Maybe we shouldn't `fmt.Sprintf` here, but rather try to depend on something like `RootSyncPermissionsName(core.RootReconcilerPrefix)` to ensure that it stays consistent if we change the naming scheme?
```suggestion
	if !strings.HasPrefix(objRef.Name, core.RootReconcilerPrefix) &&
		!strings.HasPrefix(objRef.Name, RootSyncPermissionsName(core.RootReconcilerPrefix)) {
```

tomasaschan replied:
I think I'm gonna go with this, since I prefer it myself.

sdowell replied:
Maybe we could add another const which stores this prefix value? e.g. `RootSyncClusterRoleBindingPrefix`

That way we can reuse the string and ensure consistency

tomasaschan replied:
Yeah, that's not a bad idea.

sdowell commented:
/ok-to-test

tomasaschan commented:
/retest

sdowell commented on pkg/reconcilermanager/controllers/build_names.go:
```diff
@@ -32,8 +32,11 @@ func RepoSyncPermissionsName() string {
-	return fmt.Sprintf("%s:%s", configsync.GroupName, core.RootReconcilerPrefix)
+// RootSyncPermissionsName returns root reconciler ClusterRoleBinding name.
+// e.g. configsync.gke.io:root-reconciler:my-root-sync
+func RootSyncPermissionsName(reconcilerName string) string {
```
I expect there may be a gap here during an upgrade. If the user upgrades from an older version, I expect that the old ClusterRoleBinding (the singleton `configsync.gke.io:root-reconciler`) will still include the Subjects for every `RootSync`. In this case, every pre-existing `RootSync` would still have a `ClusterRoleBinding` which grants `cluster-admin` in addition to the new `ClusterRoleBinding`  (`configsync.gke.io:root-reconciler:my-root-sync`).

tomasaschan replied:
Yes, I'm aware; see the "Open Question" in the OP. In practice, this will be quite quickly remediated _if_ there is a `RootSync` called `root-sync`, since that one will then "take over" the previously-singleton role binding and remove all the other subjects from it. The hard part to solve is the case when no such `RootSync` exists. Not sure how you prefer to handle that though.

sdowell replied:
Ah I missed that in the PR description and went straight to the design doc, sorry about that.

I'm thinking a potential approach could be:
- remove the special case for `root-reconciler` below, such that its `ClusterRoleBinding` will be named `configsync.gke.io:root-reconciler:root-sync`
- add a snippet to the Reconcile loop which deletes the legacy `ClusterRoleBinding` named `configsync.gke.io:root-reconciler`

This solution should handle the migration in all cases without user interference, even in the case that a `RootSync` named `root-sync` doesn't exist. WDYT?

tomasaschan replied:
That is cleaner, and makes the name `root-sync` less special. I like it!

tomasaschan commented on docs/design-docs/02-custom-root-reconciler-clusterrole.md:
```diff
@@ -0,0 +1,111 @@
+This role is defaulted to `cluster-admin` in order to stay backwards compatible.
+
+For the case where a single `ClusterRole` is not expressive enough to configure the
+permissions a user want, we also allow the special sentinel value `%none%`, which
```
I don't know; I don't think so. I picked it for two reasons:
1. It is a value that can never be a valid name of a `ClusterRole` (because `%` is a disallowed character), so setting it to this can never end up binding to an unintended role created later.
2. I found it more descriptive than e.g. using a single invalid character or something else along those lines.
If you don't like it we can remove that part of this PR; although I do find it appealing to have _some_ way to completely turn off the binding...

tomasaschan replied:
Another option would be to make the property pointer-valued, and differentiate between "not set" and "set to an empty string", but that seems bad for other reasons...

sdowell replied:
Another option could be to make this a struct with a nested `name` and `enabled` field. This could create a clearer API for how to disable ClusterRoleBinding, although adds more API permutations (e.g. `name=my-cluster-role, enabled=false` probably doesn't make sense).

I don't feel too strongly about it and am okay with `%none%`, my main concern is whether this is unconventional

tomasaschan replied:
I had the struct idea independently (sleeping on ideas is often effective! 😄), but I imagined it as
```
type ClusterRoleBindingOverrides struct {
  RoleName string
  Disabled   bool
}
```
This way the zero value of the struct is the current behavior (if an empty `RoleName` defaults it to `"cluster-admin"`); setting the name to something else does what this PR currently does for any value other than `%none%`, and setting `Disabled: true` disables the binding (and ignores `RoleName`, i.e. `Disabled` takes precedence).

I'll work on an update here address this (and other comments).

tomasaschan commented on docs/design-docs/02-custom-root-reconciler-clusterrole.md:
```diff
@@ -0,0 +1,111 @@
+
+## Test Plan
+
+Implementation of this feature can includes automated tests that verify the behavior.
```
There is pretty good coverage of the new and changed features here, although not among the e2e tests. In particular, see the changes to [rootsync_controller_test.go](https://github.com/GoogleContainerTools/kpt-config-sync/pull/938/files#diff-96d9758339603eab1bf5deb9d64abf4acb751b8474cd85e2cd4d790a76514291).

tomasaschan commented on pkg/api/configsync/v1beta1/resource_override.go:
```diff
@@ -92,6 +92,18 @@ type RootSyncOverrideSpec struct {
 	// +optional
 	NamespaceStrategy configsync.NamespaceStrategy `json:"namespaceStrategy,omitempty"`
+
+	// ClusterRole controls which role to bind the service account for this RootSync's
```
I followed the convention of the other doc comments, which are also `PascalCased` despite their JSON property names being `camelCased`.

Pushed a commit with the output of `make configsync-crds` too :)

tomasaschan commented on pkg/reconcilermanager/controllers/rootsync_controller.go:
```diff
@@ -779,14 +791,14 @@ func (r *RootSyncReconciler) validateRootSecret(ctx context.Context, rootSync *v
 		childCRB.OwnerReferences = nil
-		childCRB.RoleRef = rolereference("cluster-admin", "ClusterRole")
+		childCRB.RoleRef = rolereference(clusterRole, "ClusterRole")
 		childCRB.Subjects = addSubject(childCRB.Subjects, r.serviceAccountSubject(reconcilerRef))
```
Or wait, are you saying we need to ignore existing bindings here, and set it to `addSubject(nil, ...)`? That's probably correct. That case should also have a test...

sdowell replied:
Yeah since there is a 1:1 relationship between `RootSync` and `ClusterRoleBinding`, I agree that it seems correct to set this to a single subject.

This would handle the upgrade path in the case that a `RootSync` named `root-sync` exists, but requires user intervention otherwise. See my other comment for a proposal on how to handle the upgrade in the general case.

tomasaschan commented:
@sdowell I think I've addressed all your feedback now. I updated the design doc to reflect the new schema (with separate properties for the name vs enabled/disabled status), and added a test that verifies that upgrading to a version with this change will cleanly remove the old binding and add the expected new ones.

I haven't added any e2e tests, but I hope the coverage in [`rootsync_controller_test.go`](https://github.com/GoogleContainerTools/kpt-config-sync/pull/938/files#diff-96d9758339603eab1bf5deb9d64abf4acb751b8474cd85e2cd4d790a76514291) suffices.

sdowell commented on pkg/core/names.go:
```diff
@@ -28,6 +29,11 @@ const (
 
+var (
+	// RootSyncPermissionsPrefix is the prefix used for all ClusterRoleBindings granting access to Root Reconcilers
+	RootSyncPermissionsPrefix = fmt.Sprintf("%s:%s", configsync.RootSyncKind, RootReconcilerPrefix)
```
nit: this can be a `const` since it is comprised of two other consts

tomasaschan replied:
It can't, at least not if we create it with `fmt.Sprintf` (my compiler complained that it wasn't a `const` expression).

sdowell replied:
You should be able to do something like

```golang
const (
    RootSyncPermissionsPrefix = configsync.RootSyncKind + ":" + RootReconcilerPrefix
)
```

tomasaschan replied:
Ah, yeah, that might work.

sdowell commented on pkg/reconcilermanager/controllers/rootsync_controller.go:
```diff
@@ -779,14 +779,37 @@ func (r *RootSyncReconciler) validateRootSecret(ctx context.Context, rootSync *v
 		childCRB.OwnerReferences = nil
-		childCRB.RoleRef = rolereference("cluster-admin", "ClusterRole")
+		childCRB.RoleRef = rolereference(clusterRole, "ClusterRole")
 		childCRB.Subjects = addSubject(childCRB.Subjects, r.serviceAccountSubject(reconcilerRef))
```
Do we still need to support allowing multiple subjects on the CRB? It seems like we can set a single subject since the assumption is now that each CRB maps 1:1 with a RootSync. Is the idea to allow other clients to add additional subjects to the CRB?

If we leave this as-is, it seems like we might consider keeping the cleanup logic which first checks that the subjects field is empty. WDYT?

tomasaschan replied:
Hm, yeah, you're right - we should probably make this assume a single subject, and use this here
```suggestion
		childCRB.Subjects = addSubject(nil, r.serviceAccountSubject(reconcilerRef))
```
Not sure if `addSubject` is used anywhere else - if not, we could probably simplify it to e.g. not take a parameter for the existing subjects.

sdowell commented on docs/design-docs/02-custom-root-reconciler-clusterrole.md:
```diff
@@ -0,0 +1,133 @@
+spec:
+  overrides:
+    clusterRole:
+      name: my-cluster-role
```
Was the intent here to set `disabled: true`?

sdowell reviewed (commented):
Looks good! A few more comments but this is close to approval

tomasaschan commented:
@sdowell Hoping this is the one! :)

sdowell commented on docs/design-docs/02-custom-root-reconciler-clusterrole.md:
```diff
@@ -0,0 +1,133 @@
+account the reconciler is running with. However, these both put a bigger burden on the
+user in order to utilize them even for the simple use cases, which is why changing which
+role to bind to is probably the most user-friendly knob to expose. With the inclusion of
+a sentinel value to create no binding, the first of these alternatives is effectively
```
Please update this snippet to reflect that the sentinel value is no longer being used

tomasaschan replied:
Done.

sdowell commented on pkg/api/configsync/v1beta1/resource_override.go:
```diff
@@ -92,6 +92,30 @@ type RootSyncOverrideSpec struct {
+
+// ClusterRoleOverrideSpec allows to override settings for the root reconciler permissions
+type ClusterRoleOverrideSpec struct {
+	// name is the name of the ClusterRole to bind to; defaults to "cluster-admin".
```
Sorry I got a new round of thoughts after not looking at this over the weekend 😅 

I'm thinking it may be better form to use the following:
```
// +kubebuilder:default:=cluster-admin
```
This would clearly document the default value on the field as well as provide the default behavior at the API level.

---

Additional consideration:

With this realization I'm also considering that we could go back to a single string field and treat it as "disabled" if the user sets the value to `""` (empty string). WDYT?

sdowell replied:
Correction: it looks like the default value only applies if the nested struct (in this case `override`) is set. So that means if the user has an empty override field, the default value would not apply (and `namespaceStrategy` would be `""`). Given this, it looks like we would still need the defaulting logic in the controller and wouldn't be able to rely on an empty string check.

tomasaschan commented on docs/design-docs/02-custom-root-reconciler-clusterrole.md:
```diff
@@ -0,0 +1,133 @@
+account the reconciler is running with. However, these both put a bigger burden on the
+user in order to utilize them even for the simple use cases, which is why changing which
+role to bind to is probably the most user-friendly knob to expose. With the inclusion of
+a sentinel value to create no binding, the first of these alternatives is effectively
+supported, if through a slightly worse API. 
```
```suggestion
a switch to turn off creating the binding entirely, the first of these alternatives is effectively
supported as well, and the API supports extending for other use cases in the future. 
```

karlkfi commented:
Circling back to this, I think we may need to extend this to support RepoSyncs too, and also multiple bindings, not just one. And if we're quick about it, we can make reverse incompatible changes before the 1.17.0 release.

I'm thinking something like:

```
  override:
    roleRefs:
    - kind: ClusterRole
      name: crd-admin
    - kind: Role
      name: admin
      namespace: example
```

This ouwld hav emultiple advantages:
1. This would allow 0-many bindings.
2. Combining multiple roles means fewer use cases where you need to provision a custom role before using it.
3. You could grant permission to provision to multiple specific namespaces.
4. It would make managing RepoSyncs easier, because you don't need to specify the RoleBinding or the dependency, which is a common gotcha when deleting, especially with deletion-propagation.
5. It would make the two APIs more similar.

Would this satisfy your needs @tomasaschan?
If you can manage multiple, do you still need to be able to disable the cluster-admin on RootSyncs?

tomasaschan commented:
@karlkfi As long as adding this override means the reconciler isn't granted cluster-admin too, I think this would work for us.

We don't have an explicit use case for turning the binding off entirely, but I can imagine two:

1. The bindings you need are complicated enough that you cannot express them in a single cluster role. That is solved by your proposal, and arguably better since the user doesn't have to create the bindings manually.

2. You're using a different authz mechanism than k8s RBAC, eg WLI + GCP IAM, and don't want to create any RBAC permissions that interfere. I think this could be worked around by binding to a role with no permissions, and might be uncommon enough in practice that a workaround like that, hacky as it is, is OK.

In other words: go ahead!

tomasaschan commented:
@karlkfi @sdowell Just to clarify: will your team do this work? We still need this feature, and want to have it available in the shape it's going to take when released as soon as we can - so I'll do it if I have to, but I have other things I'd rather spend time on...

karlkfi commented:
@sdowell has been working on it: https://github.com/GoogleContainerTools/kpt-config-sync/pull/977

janetkuo commented on docs/design-docs/02-custom-root-reconciler-clusterrole.md:
```diff
@@ -0,0 +1,133 @@
+spec:
+  overrides:
+    clusterRole:
+      disabled: true
```
Whoever has access to RootSync has access to grant themselves any permission they want by specifying the right cluster role. This has the problem of privilege escalation. However, given that RootSync has cluster-admin access before this change, this change isn't that dangerous, but could be surprising to admins when they grant limited access but also grant RootSync access to someone else. @tomasaschan what do you think about this caveat? Ideally the toggle should live outside of RootSync, but it's less of a problem for RootSync compared to RepoSync API.

tomasaschan replied:
In the setup we run which motivates this feature this won't be an issue - we do not intend to grant access to create new syncs or update existing ones. In fact, the whole point is to keep all syncs in a repo we control, so that we can review the setup and ensure permissions are what they should be, while allowing them to sync config from repos we don't. In other words, this feature is what allows us to _prevent_ privilege escalation.

So I don't think there's a privilege escalation vector here that wasn't here already. If you can create or edit syncs, you can sync in config from an arbitrary source, and grant yourself whatever permissions you want.

As for RepoSyncs, in addition to all of the above, they will not be able to create cluster scoped resources, so any escalation will be even less limited.

janetkuo replied:
The privilege escalation problem is more on the RepoSync given that it has no permission to do anything by default. 

>  If you can create or edit syncs, you can sync in config from an arbitrary source, and grant yourself whatever permissions you want.

This isn't true for RepoSync. RepoSync editors don't have access to modify the RBAC rolebindings granted to the namespace reconciler SA.
